nixos/modules/services/web-apps/immich.nix

```
@@ -0,0 +1,575 @@
+    )}
+
+    ${optionalString serverCfg.typesense.enable ''
+      export TYPESENSE_API_KEY=$(cat ${serverCfg.typesense.apiKeyFile})
+    ''}
+
+    exec ${program}
+  '';
+
+  commonServiceConfig = {
+    Restart = "on-failure";
+
+    # Hardening
+    CapabilityBoundingSet = "";
+    LockPersonality = true;
+    MemoryDenyWriteExecute = true;
+    NoNewPrivileges = true;
+    PrivateUsers = true;
+    PrivateTmp = true;
+    PrivateDevices = true;
+    PrivateMounts = true;
+    ProtectClock = true;
+    ProtectControlGroups = true;
+    ProtectHome = true;
+    ProtectHostname = true;
+    ProtectKernelLogs = true;
+    ProtectKernelModules = true;
+    ProtectKernelTunables = true;
+    ProtectProc = "invisible";
+    ProcSubset = "pid";
+    # Would re-mount paths ignored by temporary root
+    # TODO ProtectSystem = "strict";
+    RemoveIPC = true;
+    RestrictAddressFamilies = [
+      "AF_INET"
+      "AF_INET6"
+      "AF_UNIX"
+    ];
+    RestrictNamespaces = true;
+    RestrictRealtime = true;
+    RestrictSUIDSGID = true;
+    SystemCallArchitectures = "native";
+    SystemCallFilter = [
+      "@system-service"
+      "~@privileged"
+      "@pkey"
+    ];
+    UMask = "0077";
+  };
+
+  serverServiceConfig = {
+    DynamicUser = true;
+    User = "immich";
+    Group = "immich";
+    SupplementaryGroups = optional enableRedis redisServerCfg.user;
+
+    StateDirectory = "immich";
+    StateDirectoryMode = "0750";
+    WorkingDirectory = "/var/lib/immich";
+
+    MemoryDenyWriteExecute = false; # nodejs requires this.
+    EnvironmentFile = mkIf (serverCfg.environmentFile != null) serverCfg.environmentFile;
+
+    TemporaryFileSystem = "/:ro";
+    BindReadOnlyPaths = [
+      "/nix/store"
+      "-/etc/resolv.conf"
+      "-/etc/nsswitch.conf"
+      "-/etc/hosts"
+      "-/etc/localtime"
+      "-/run/postgresql"
+    ] ++ optional enableRedis redisServerCfg.unixSocket;
+  };
+in {
+  options.services.immich = {
+    enable = mkEnableOption "immich" // {
+      description = ''
+        Enables immich which consists of a backend server, microservices,
+        machine-learning and web ui. You can disable or reconfigure components
+        individually using the subsections.
+      '';
+    };
+
+    package = mkPackageOption pkgs "immich" {};
+
+    server = {
+      mediaDir = mkOption {
+        type = types.str;
+        default = "/var/lib/immich/media";
+        description = "Directory used to store media files.";
+      };
+
+      backend = {
+        enable = mkEnableOption "immich backend server" // {
+          default = true;
+        };
+        port = mkOption {
+          type = types.port;
+          default = 3001;
+          description = "Port to bind to.";
+        };
+
+        openFirewall = mkOption {
+          default = false;
+          type = types.bool;
+          description = "Whether to open the firewall for the specified port.";
+        };
+
+        extraConfig = mkOption {
```

```suggestion
        settings = mkOption {
```

https://github.com/NixOS/rfcs/pull/42